GUVNOR-2173 - Asset Oracle: Change Impact prediction

mrietveld opened the pull request:
Change Impact, Part I: building a reference infrastructure that we can query.

For an explanation of the changes, see https://github.com/droolsjbpm/kie-wb-common/pull/414

PR set:
- https://github.com/uberfire/uberfire/pull/455
- https://github.com/droolsjbpm/kie-wb-common/pull/414
- https://github.com/droolsjbpm/drools-wb/pull/257
- https://github.com/droolsjbpm/jbpm-designer/pull/375
- https://github.com/droolsjbpm/jbpm-form-modeler/pull/63

Also, this PR depends on https://github.com/droolsjbpm/jbpm/pull/511

manstis commented on jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/BPMN2FileIndexer.java:
```diff
@@ -0,0 +1,162 @@
+package org.jbpm.designer.server.indexing;
```
License

manstis commented on jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/BPMN2FileIndexer.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Override
+    public KObjectKey toKObjectKey( Path path ) {
```
I thought I read somewhere that this method should not be overridden and ideally it'd have been ```final```?

manstis commented on jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/ProcessDefintionIndexVisitor.java:
```diff
@@ -0,0 +1,73 @@
+package org.jbpm.designer.server.indexing;
```
License

manstis commented on jbpm-designer-backend/src/main/java/org/jbpm/designer/server/indexing/ProcessIndexVisitor.java:
```diff
@@ -0,0 +1,83 @@
+package org.jbpm.designer.server.indexing;
```
License

manstis commented on jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java:
```diff
@@ -177,6 +182,108 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+        Collections.sort( otherRuleFlowGroupNames );
+
+        // divider between asset project/branch ruleflowgroup names and all ruleflowgroup names
+        ruleFlowGroupNames.add("********");
```
I can't see that the old way had any demarcation between the two sets of Rule Flow Group Names. If you're adding the distinction I'd assume something somewhere client-side would need tweaking to handle it too?

mrietveld replied:
This was a suggestion by @tsurdilo -- Tiho, could you +1 this change?

manstis commented on jbpm-designer-backend/src/main/java/org/jbpm/designer/web/server/CalledElementServlet.java:
```diff
@@ -177,6 +182,108 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+        Collections.sort( otherDataTypeNames );
+
+        // divider between asset project/branch ruleflowgroup names and all ruleflowgroup names
+        dataTypeNames.add("********");
```
Same here.

mrietveld replied:
@tsurdilo +1, please? :)

mrietveld commented:
@tsurdilo  Would you mind also reviewing this? Thanks! It will help to read droolsjbpm/kie-wb-common#414 first.

mrietveld commented:
Fixed missing licenses.

mrietveld commented:
jenkins retest this

tsurdilo commented:
@mrietveld could i pls have couple days to test this locally to make sure there is no impact on ui? im in middle of bigger pr now so pls give me just couple days or if @jlindop has some time now he can test as well. thanks.

mrietveld commented:
@tsurdilo no problem.

mrietveld commented:
jenkins retest this

(*really* fixed the non-public method/errai marshalling problems.. )

manstis commented:
@mrietveld This is still red.

mrietveld commented:
Mr. Jenkins, the please of the retest would be green. Yes. ;)

mrietveld commented:
There was once a Jenkins that ate tests. "Please, may I have some more?" it would ask and then retest.

jlindop commented:
@mrietveld   This builds now, which is good, but I haven't been able to test it.
I've re-built all the repo's which still have unmerged commits for GUVNOR-2173 and built kie-wb-distributions/kie-wb, then deployed the wildfl10 war file to wildfly10. The war fails to start with this error:

16:17:16,835 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-4) MSC000001: Failed to start service jboss.deployment.unit."kie-wb-7.0.0-SNAPSHOT-wildfly10.war".WeldStartService: org.jboss.msc.service.StartException in service jboss.deployment.unit."kie-wb-7.0.0-SNAPSHOT-wildfly10.war".WeldStartService: Failed to start service
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1904)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type DefinitionService with qualifiers @Default
  at injection point [BackedAnnotatedField] @Inject protected org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service
  at org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service(BPMN2FileIndexer.java:0)

	at org.jboss.weld.bootstrap.Validator.validateInjectionPointForDeploymentProblems(Validator.java:359)
	at org.jboss.weld.bootstrap.Validator.validateInjectionPoint(Validator.java:281)
	at org.jboss.weld.bootstrap.Validator.validateGeneralBean(Validator.java:134)
	at org.jboss.weld.bootstrap.Validator.validateRIBean(Validator.java:155)
	at org.jboss.weld.bootstrap.Validator.validateBean(Validator.java:518)
	at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:68)
	at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:66)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:63)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:56)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)

16:17:16,844 ERROR [org.jboss.as.controller.management-operation] (Controller Boot Thread) WFLYCTL0013: Operation ("deploy") failed - address: ([("deployment" => "kie-wb-7.0.0-SNAPSHOT-wildfly10.war")]) - failure description: {"WFLYCTL0080: Failed services" => {"jboss.deployment.unit.\"kie-wb-7.0.0-SNAPSHOT-wildfly10.war\".WeldStartService" => "org.jboss.msc.service.StartException in service jboss.deployment.unit.\"kie-wb-7.0.0-SNAPSHOT-wildfly10.war\".WeldStartService: Failed to start service
    Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type DefinitionService with qualifiers @Default
  at injection point [BackedAnnotatedField] @Inject protected org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service
  at org.jbpm.designer.server.indexing.BPMN2FileIndexer.bpmn2Service(BPMN2FileIndexer.java:0)
"}}

mrietveld commented:
@jlindop Thanks, looking into it.

mrietveld commented:
@jlindop Fixed. The problem was that the `jbpm-services-cdi` module, which produces the `DefinitionService` CDI impl, was not included in either `jbpm-designer-standalone` or the workbench.

mrietveld commented:
Argh.. fixing. `jbpm-kie-services` includes `dashbuilder` jars, which mess up the build..

mrietveld commented:
@jlindop Argh, needs another PR to fix this.

csadilek commented:
@cristianonicolai as I will be off for a couple of days, can you merge this PR series once this one as a plus one and is green? Thanks.

cristianonicolai commented:
@csadilek will do ;)

jlindop commented:
@jomarko @tsurdilo  Will one of you test this when it's ready to test (Marco's working on it)? See above for the set of dependent PR's. I'm on PTO next week, and @mrietveld  is leaving on Aug31st, so we should try to merge it as soon as we can. Thanks, Jeremy.

jomarko commented:
@jlindop I can't promise, but if I will have a moment I can try it. I will contact @tsurdilo when the PR will be ready and I will be out of time.

mrietveld commented:
@jomarko @tsurdilo (and @mswiderski ?). This is done. If you have any questions, please let me know. 

( @manstis Once this is done, the rest can be merged as well? )

mswiderski commented:
looks very good to me - from the process information capturing point of view. Thanks @mrietveld for implementing the listener in designer. I do believe that is the cleanest solution even though it does duplicate tiny bit of code.

mrietveld commented:
@jomarko or @tsurdilo, would either of you have some time to review this? I have 3 other PR's waiting on this. 

Thanks!

jomarko commented:
Here are my comments:

- `DesignerProcessDataEventListener` I did not find tests for this class and it seems to be as worth to be tested. Next there is typo error in variable `functionImpot`. It should be `functionImpot`. And the last question to this class. There is used is code `StringUtils.isEmpty()`. I looked at implementation of this and I think that it evaluates to false for cases like this `StringUtils.isEmpty(' ')`. Is it fine?

- `DesignerFindDataTypesQuery` and `FindRuleFlowNamesQuery`. In these classes I am wondering about methods `toQuery()` and `validateTerms()`. 

I think that in `DesignerFindDataTypesQuery.toQuery(terms)` is redundant check if `terms` are not null and empty.

Shouldn't be also in `DesignerFindDataTypesQuery.validateTerms(terms)` code that ensures that required term was found? Please compare `DesignerFindDataTypesQuery.validateTerms(terms)` with `FindRuleFlowNamesQuery.validateTerms(terms)` to see my point.

mrietveld commented:
@jomarko Thanks!

- Tests: the test for the `DesignerProcessDataEventListener` class is this class: `BPMN2FileIndexerIndexingTest`. 
- `functionImpot`: will fix
- redundant check: I'll remove that
- `validateTerms`: will fix

Thanks.

jomarko commented:
@mrietveld thanks, and the last thing. Is it fine if `StringUtils.isEmpty(' ')` evaluates to false in `DesignerProcessDataEventListener`?

mrietveld commented:
@jomarko Yes, that's correct (it should evaluate to false and *not* to true!)

mrietveld commented:
@jomarko Thanks for the feedback -- I've fixed the `validateTerms` and `toQuery` methods, and also added a test for the `validateTerms` method. 

(The `toQuery` is tested in the indexing test).

mrietveld commented:
jenkins please retest

jomarko commented:
+1
